syntax = "proto3";

package loadbalancer;

// Smart Load Balancer Service v3.0
service LoadBalancer {
    // Client registration with smart model assignment
    rpc RegisterClient(ClientInfo) returns (RegistrationResponse);
    
    // Process AI request
    rpc ProcessAIRequest(AIRequest) returns (AIResponse);
    
    // Get processing status (for progress tracking)
    rpc GetProcessingStatus(StatusRequest) returns (StatusResponse);
    
    // Health check
    rpc HealthCheck(Empty) returns (HealthResponse);
    
    // Get available models (for client verification)
    rpc GetAvailableModels(Empty) returns (AvailableModelsResponse);
    
    // Reassign models (for dynamic rebalancing)
    rpc ReassignModels(Empty) returns (ReassignmentResponse);
    
    // RAG operations
    rpc AddDocument(AddDocumentRequest) returns (AddDocumentResponse);
    rpc SearchDocuments(SearchDocumentsRequest) returns (SearchDocumentsResponse);
    
    // Chat history operations
    rpc CreateChatSession(CreateChatSessionRequest) returns (CreateChatSessionResponse);
    rpc GetChatSessions(GetChatSessionsRequest) returns (GetChatSessionsResponse);
    rpc GetChatHistory(GetChatHistoryRequest) returns (GetChatHistoryResponse);
}

// Empty message
message Empty {}

// Client information
message ClientInfo {
    string client_id = 1;
    string hostname = 2;
    string ip_address = 3;
    SystemSpecs specs = 4;
}

// System specifications
message SystemSpecs {
    int32 cpu_cores = 1;
    float cpu_frequency_ghz = 2;
    int64 ram_gb = 3;
    string gpu_info = 4;
    float gpu_memory_gb = 5;
    string os_info = 6;
    float performance_score = 7;
}

// Registration response
message RegistrationResponse {
    bool success = 1;
    string message = 2;
    string assigned_model = 3;
    ModelInfo model_info = 4;
    int32 total_clients = 5;
    int32 client_group = 6;
}

// Model information
message ModelInfo {
    string name = 1;
    int64 parameters = 2;
    float size_gb = 3;
    int32 complexity_score = 4;
}

// Available models response
message AvailableModelsResponse {
    repeated ModelInfo models = 1;
    int32 total_models = 2;
}

// Reassignment response
message ReassignmentResponse {
    bool success = 1;
    string message = 2;
    repeated ClientAssignment new_assignments = 3;
}

// Client assignment
message ClientAssignment {
    string client_id = 1;
    string assigned_model = 2;
    int32 group_number = 3;
}

// AI Request
message AIRequest {
    string request_id = 1;
    string prompt = 2;
    string assigned_model = 3;
    int64 timestamp = 4;
    repeated ImageData images = 5;  // Support for multimodal input
    string chat_session_id = 6;  // For chat history tracking
    repeated string context_documents = 7;  // RAG context
}

// AI Response
message AIResponse {
    string request_id = 1;
    bool success = 2;
    string response_text = 3;
    float processing_time = 4;
    string client_id = 5;
    string model_used = 6;
    int64 timestamp = 7;
}

// Status request for progress tracking
message StatusRequest {
    string request_id = 1;
    string client_id = 2;
}

// Status response for progress tracking
message StatusResponse {
    string request_id = 1;
    string client_id = 2;
    ProcessingStatus status = 3;
    float progress_percentage = 4;
    string current_step = 5;
    int64 estimated_remaining_seconds = 6;
}

// Processing status enum
enum ProcessingStatus {
    IDLE = 0;
    STARTING = 1;
    PROCESSING = 2;
    FINALIZING = 3;
    COMPLETED = 4;
    ERROR = 5;
}

// Health check response
message HealthResponse {
    bool healthy = 1;
    string message = 2;
    int32 connected_clients = 3;
    int32 active_models = 4;
}

// Image data for multimodal support
message ImageData {
    string image_base64 = 1;  // Base64 encoded image
    string mime_type = 2;  // e.g., "image/jpeg", "image/png"
    string filename = 3;  // Optional filename
}

// Document for RAG
message Document {
    string doc_id = 1;
    string content = 2;
    string title = 3;
    map<string, string> metadata = 4;
    int64 timestamp = 5;
}

// Chat message
message ChatMessage {
    string message_id = 1;
    string session_id = 2;
    string role = 3;  // "user" or "assistant"
    string content = 4;
    int64 timestamp = 5;
    repeated ImageData images = 6;
}

// Chat session
message ChatSession {
    string session_id = 1;
    string title = 2;
    repeated ChatMessage messages = 3;
    int64 created_at = 4;
    int64 updated_at = 5;
}

// RAG operations
message AddDocumentRequest {
    Document document = 1;
}

message AddDocumentResponse {
    bool success = 1;
    string message = 2;
    string doc_id = 3;
}

message SearchDocumentsRequest {
    string query = 1;
    int32 top_k = 2;  // Number of results to return
}

message SearchDocumentsResponse {
    repeated Document documents = 1;
    repeated float scores = 2;
}

// Chat history operations
message GetChatSessionsRequest {
    int32 limit = 1;
}

message GetChatSessionsResponse {
    repeated ChatSession sessions = 1;
}

message GetChatHistoryRequest {
    string session_id = 1;
}

message GetChatHistoryResponse {
    ChatSession session = 1;
    bool success = 2;
}

message CreateChatSessionRequest {
    string title = 1;
}

message CreateChatSessionResponse {
    string session_id = 1;
    bool success = 2;
}